# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("/Users/admin/Desktop/IRONHACK/2.WEEK/LABS_2/lab-dw-data-structuring-and-combining/marketing_customer_analysis_clean.csv")

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 27 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   unnamed:_0                     10910 non-null  int64  
 1   customer                       10910 non-null  object 
 2   state                          10910 non-null  object 
 3   customer_lifetime_value        10910 non-null  float64
 4   response                       10910 non-null  object 
 5   coverage                       10910 non-null  object 
 6   education                      10910 non-null  object 
 7   effective_to_date              10910 non-null  object 
 8   employmentstatus               10910 non-null  object 
 9   gender                         10910 non-null  object 
 10  income                         10910 non-null  int64  
 11  location_code                  10910 non-null  object 
 12  marital_status                 10910 non-null 

In [4]:
df.head()

unnamed:_0 customer       state  customer_lifetime_value response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   coverage education effective_to_date employmentstatus gender  ...  \
0     Basic   College        2011-02-18         Employed      M  ...   
1     Basic   College        2011-01-18       Unemployed      F  ...   
2     Basic  Bachelor        2011-02-10         Employed      M  ...   
3  Extended   College        2011-01-11         Employed      M  ...   
4   Premium  Bachelor        2011-01-17    Medical Leave      F  ...   

   number_of_policies     policy_type        policy  renew_offer_type  \
0                   9  Corporate Auto  Corporate L3            Offer3   
1                   1   Personal Auto   Personal L3            Offer4   
2                   2   Personal Auto   Personal L3            Offer3   
3                   2  Corporate Auto  Corporate L3            Offer2   
4                   7   Personal Auto   Personal L2            Offer1   

   sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0          Agent          292.800000  Four-Door Car       Medsize   
1    Call Center          744.924331  Four-Door Car       Medsize   
2    Call Center          480.000000            SUV       Medsize   
3         Branch          484.013411  Four-Door Car       Medsize   
4         Branch          707.925645  Four-Door Car       Medsize   

  vehicle_type month  
0            A     2  
1            A     1  
2            A     2  
3            A     1  
4            A     1  

[5 rows x 27 columns]

In [5]:
filter = df.loc[(df['total_claim_amount'] > 1000) & (df['response'] == 'Yes')]
filter

unnamed:_0 customer       state  customer_lifetime_value response  \
189           189  OK31456  California             11009.130490      Yes   
236           236  YJ16163      Oregon             11009.130490      Yes   
419           419  GW43195      Oregon             25807.063000      Yes   
442           442  IP94270     Arizona             13736.132500      Yes   
587           587  FJ28407  California              5619.689084      Yes   
...           ...      ...         ...                      ...      ...   
10351       10351  FN44127      Oregon              3508.569533      Yes   
10373       10373  XZ64172      Oregon             10963.957230      Yes   
10487       10487  IX60941      Oregon              3508.569533      Yes   
10565       10565  QO62792      Oregon              7840.165778      Yes   
10708       10708  CK39096      Oregon              5619.689084      Yes   

       coverage             education effective_to_date employmentstatus  \
189     Premium              Bachelor        2011-01-24         Employed   
236     Premium              Bachelor        2011-01-24         Employed   
419    Extended               College        2011-02-13         Employed   
442     Premium                Master        2011-02-13         Disabled   
587     Premium  High School or Below        2011-01-26       Unemployed   
...         ...                   ...               ...              ...   
10351  Extended               College        2011-01-05    Medical Leave   
10373   Premium  High School or Below        2011-02-08         Employed   
10487  Extended               College        2011-01-05    Medical Leave   
10565  Extended               College        2011-01-14         Employed   
10708   Premium  High School or Below        2011-01-26       Unemployed   

      gender  ...  number_of_policies     policy_type        policy  \
189        F  ...                   1  Corporate Auto  Corporate L3   
236        F  ...                   1    Special Auto    Special L3   
419        F  ...                   2   Personal Auto   Personal L2   
442        F  ...                   8   Personal Auto   Personal L2   
587        M  ...                   1   Personal Auto   Personal L1   
...      ...  ...                 ...             ...           ...   
10351      M  ...                   1   Personal Auto   Personal L2   
10373      M  ...                   1  Corporate Auto  Corporate L2   
10487      M  ...                   1   Personal Auto   Personal L3   
10565      M  ...                   1   Personal Auto   Personal L3   
10708      M  ...                   1   Personal Auto   Personal L3   

       renew_offer_type  sales_channel  total_claim_amount  vehicle_class  \
189              Offer2          Agent         1358.400000     Luxury Car   
236              Offer2          Agent         1358.400000     Luxury Car   
419              Offer1         Branch         1027.200000     Luxury Car   
442              Offer1            Web         1261.319869            SUV   
587              Offer2            Web         1027.000029            SUV   
...                 ...            ...                 ...            ...   
10351            Offer2         Branch         1176.278800  Four-Door Car   
10373            Offer1          Agent         1324.800000     Luxury SUV   
10487            Offer2         Branch         1176.278800  Four-Door Car   
10565            Offer2          Agent         1008.000000  Four-Door Car   
10708            Offer2            Web         1027.000029            SUV   

       vehicle_size vehicle_type month  
189         Medsize            A     1  
236         Medsize            A     1  
419           Small            A     2  
442         Medsize            A     2  
587         Medsize            A     1  
...             ...          ...   ...  
10351         Small            A     1  
10373       Medsize            A     2  
10487         Small            A  

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

In [6]:
avg_claim_amount = df.loc[df['response'] == 'Yes'].groupby(['policy_type', 'gender']) ['total_claim_amount'].mean()
avg_claim_amount

policy_type     gender
Corporate Auto  F         433.738499
                M         408.582459
Personal Auto   F         452.965929
                M         457.010178
Special Auto    F         453.280164
                M         429.527942
Name: total_claim_amount, dtype: float64

**Corporate Auto:**
- For female customers, the average total claim amount is approximately 433.74, whereas for male customers, it is slightly lower, around $408.58. This suggests that claims made by female customers in the Corporate Auto category tend to be slightly higher on average than those made by male customers.
  
**Personal Auto:**
- In the Personal Auto category, female customers have a slightly lower average claim amount, around 452.97, compared to male customers, who have an average claim amount of approximately 457.01. The difference between genders is minimal in this category.
  
**Special Auto:**
- For the Special Auto category, we observe a similar pattern to Corporate Auto, where claims made by female customers have a slightly higher average, around 453.28, compared to male customers, who have an average claim amount of approximately 429.53.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [7]:
state_customers = df.groupby('state').size().reset_index(name='total_customers')
states_with_more_than_500_customers = state_customers[state_customers['total_customers'] > 500]
print(states_with_more_than_500_customers)

        state  total_customers
0     Arizona             1937
1  California             4183
2      Nevada              993
3      Oregon             2909
4  Washington              888


4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [8]:
customer_lifetime_stats = df.groupby(['education', 'gender'])['customer_lifetime_value'].agg(['max', 'min', 'median']).reset_index()
print(customer_lifetime_stats)

              education gender          max          min       median
0              Bachelor      F  73225.95652  1904.000852  5640.505303
1              Bachelor      M  67907.27050  1898.007675  5548.031892
2               College      F  61850.18803  1898.683686  5623.611187
3               College      M  61134.68307  1918.119700  6005.847375
4                Doctor      F  44856.11397  2395.570000  5332.462694
5                Doctor      M  32677.34284  2267.604038  5577.669457
6  High School or Below      F  55277.44589  2144.921535  6039.553187
7  High School or Below      M  83325.38119  1940.981221  6286.731006
8                Master      F  51016.06704  2417.777032  5729.855012
9                Master      M  50568.25912  2272.307310  5579.099207


- **Bachelor Degree Holders:** Female bachelor degree holders tend to have slightly higher CLVs compared to male bachelor degree holders, with higher maximum and median values.
- **College Degree Holders:** There is a slight difference in the maximum CLV between female and male college degree holders, with females having a higher maximum CLV. However, male college degree holders have a higher median CLV compared to females.
- **Doctor Degree Holders:** Female doctor degree holders have a higher maximum CLV compared to males. However, males have a higher median CLV.
- **High School or Below Education Level:**  Males with a high school diploma or below education level have a significantly higher maximum CLV compared to females. However, females have a higher median CLV.
- **Master's Degree Holders:** There is a slight difference in the maximum CLV between female and male master's degree holders, with females having a slightly higher maximum CLV. However, males have a slightly higher median CLV.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [ ]:
# your code goes here